### PR TITLE
Find Agent by Name 

### PR DESCRIPTION
**Pull request by meg-krieg**
# Summary of Changes

Resolves issue #1920 to add look up capability in context for finding agent by prototype name. 

# Testing and Validation

- [x] I have read the [Contributing to Cyclus](https://fuelcycle.org/kernel/contributing_to_cyclus.html) guide
- [x] I have compiled and run the code locally
- [x] I have added or updated relevant tests
- [x] I have added documentation for new or changed features
- [x] This code follows the style guide
- [x] I have updated the changelog

Reviewers, please refer to the Cyclus [Guide for Reviewers](https://fuelcycle.org/kernel/pr_review.html).

### CODE REVIEW

File: src/context.cc [hunk 1]
```diff
@@ -116,6 +116,16 @@ void Context::DelAgent(Agent* m) {
...
 }
 
+Agent* Context::GetAgentByName(std::string name){
+  std::set<Agent*>::iterator it;
+  for (it = agent_list_.begin(); it != agent_list_.end(); ++it){
+    if((*it)->prototype() == name){
+      return (*it);
+    }
+  }
+  cyclus::Warn<KEY_WARNING>("Agent set does not contain " + name + " name");
```
Comment (gonuke) [inline src/context.cc:126]:
```suggestion
  cyclus::Warn<KEY_WARNING>("Context set does not contain agent with name: " + name);
```

File: tests/context_tests.cc [hunk 2]
```diff
@@ -177,4 +177,27 @@ TEST_F(ContextTests, GetAgentList) {
...
+  Recorder rec;
+  Context* ctx = new Context(&ti, &rec);
+
+  Agent* agent1 = new DonutShop(ctx, "chocolate");
+  Agent* agent2 = new DonutShop(ctx, "chocolate");
+  Agent* agent3 = new DonutShop(ctx, "chocolate");
+
+  agent1->prototype("dunkin donuts");
+  agent2->prototype("krispy kreme");
+  agent3->prototype("voodoo donut");
```
Comment (gonuke) [inline tests/context_tests.cc:193]: I recommend setting this test up in the same way as `CreateAgents`.

File: src/context.cc [hunk 3]
```diff
@@ -116,6 +116,16 @@ void Context::DelAgent(Agent* m) {
   }
 }
 
+Agent* Context::GetAgentByName(std::string name){
```
Comment (gonuke) [inline src/context.cc:119]: I realize that we don't ever actually name agents, but only name their prototypes,so maybe this should be:
```suggestion
Agent* Context::GetAgentByPrototypeName(std::string name){
```
Or
```suggestion
Agent* Context::GetAgentPrototypeByName(std::string name){
```
?

File: tests/context_tests.cc [hunk 4]
```diff
@@ -177,4 +177,27 @@ TEST_F(ContextTests, GetAgentList) {
...
+
+  ASSERT_NO_THROW(ctx->GetAgentByName("dunkin donuts"));
+
+  cyclus::warn_as_error = true;
+  EXPECT_THROW(ctx->GetAgentByName("donut shop"), cyclus::KeyError);
+
+  cyclus::warn_as_error = false;
+               
+  delete ctx;
```
Comment (dean-krueger) [inline tests/context_tests.cc:202]: Two thoughts here:

1. Is there a way to check that the function actually gives you the agent (It looks like right now you only check that it doesn't throw when you ask for it)
2. I always ALWAYS forget how this works, but in simulations sometimes we make a facility and call it something like "UraniumMine" and then deploy 50 of them. If I do GetAgentByName(UraniumMine)` this gives the like UraniumMine that's being copied 50 times and not a list of all the copies? If so, is that what we want? If not, what happens if it sees more than one `UraniumMine`? Can you make two facilities with the same name but different implementations? There may be a little bit of edge-case testing that is needed.

I hope that second bullet point makes sense?

**Review by dean-krueger — COMMENTED** [review]
Thanks @meg-krieg ! This looks really good to me (though many times I have thought that and then a more experienced reviewer brings up some good points). I left a comment about some testing. I have no idea how possible some of that is, but I think it's worthwhile to think a bit about edge cases and maybe try to do deeper testing than just not throwing. I'm going to hit the "Comment" button so that if what I was saying about the tests was nonsense you don't get blocked.

**Review by gonuke — CHANGES_REQUESTED** [review]
Thanks @meg-krieg - I started this review and then went to leave comments in the issue as well on the high level goal here...  There are a few suggestions here, assuming we move forward in some way.

File: tests/context_tests.cc (hunk 4 above)
Comment (gonuke) [reply tests/context_tests.cc:202]: @dean-krueger this second point occurred to me also and I went to the issue (#1920)  to discuss it further since I think the issue (by me) was not well-posed.
